### PR TITLE
chore: make CPython 3.12 default CI version

### LINKED ISSUE #2196: Make Python 3.12 the default CI version
This is a meta issue tracking full 3.12 support.

- [x] Basic 3.12 support: #2176.
- [x] Migrate all applicable CI gates to use 3.12.
- [x] Update `CONTRIBUTING.md` noting it is fine to use 3.12 for development.
- [x] Make the contributor's checklist pass on 3.12: #2179.
- [x] Address all deprecations in 3.12: #2178, #2066.

### PR DESCRIPTION
**Pull request by vytas7**
Fixes #2196

### CODE REVIEW

**Comment by vytas7** [comment]
I'll update branch rules/expected checks if this looks good otherwise.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -37,7 +37,7 @@ $ pip install -U blue
 $ blue .
 ```
 
-You can check all this by running ``tox`` from within the Falcon project directory. Your environment must be based on CPython 3.8, 3.10 or 3.11:
+You can check all this by running ``tox`` from within the Falcon project directory. Your environment must be based on CPython 3.8, 3.10, 3.11 or 3.12:
```
Comment (CaselIT) [inline CONTRIBUTING.md:40]: Maybe it's ok to just use 3.8-3.12

**Review by CaselIT — APPROVED** [review]

File: CONTRIBUTING.md (hunk 1 above)
Comment (vytas7) [reply CONTRIBUTING.md:40]: 3.9 doesn't work well (explained below), so I think it is better to keep it clear. Moreover, it was like that before this PR.

Comment (CaselIT) [reply CONTRIBUTING.md:40]: oh, forgot about that 3.9 issue

**Comment by vytas7** [comment]
Updated the checks to make the new ones `Required` except `py37` as we'll probably be removing it.

**Pull request merged**